data_structures/src/chain/priority.rs

```diff
@@ -0,0 +1,1382 @@
+        target_minutes: [u16; 5],
+        seconds_per_epoch: u16,
+    ) -> PrioritiesEstimate
+    where
+        I: IntoIterator<Item = &'a Priorities>,
+        I::IntoIter: Clone + ExactSizeIterator,
+    {
+        // Make the priorities argument an iterator (if it was not already) and measure its length.
+        let priorities = priorities.into_iter();
+        let priorities_count = priorities.len() as f64;
+
+        // Set the number of buckets used to ease the frequency counting
+        let buckets_count = 50.0;
+
+        // Create counters for measuring frequency of priorities separately for DRTs and VTTs.
+        let mut drt_counter = counter::Counter::<u64>::new();
+        let mut vtt_counter = counter::Counter::<u64>::new();
+
+        // This is a first pass over the priorities in the engine, just to find out the absolute
+        // minimum and maximum among all the lowest priorities, i.e. what was the priority for the
+        // less prioritized transaction in the blocks with the lowest and highest priority
+        // requirements.
+        let (drt_lowest_absolute, drt_highest_absolute, vtt_lowest_absolute, vtt_highest_absolute) =
+            priorities.clone().fold(
+                (f64::MAX, 0.0f64, f64::MAX, 0.0f64),
+                |(drt_lowest, drt_highest, vtt_lowest, vtt_highest), priorities| {
+                    let drt_min = priorities
+                        .drt_lowest
+                        .unwrap_or(priorities.drt_highest)
+                        .as_f64();
+                    let vtt_min = priorities
+                        .vtt_lowest
+                        .unwrap_or(priorities.vtt_highest)
+                        .as_f64();
+
+                    (
+                        drt_lowest.min(drt_min),
+                        drt_highest.max(drt_min),
+                        vtt_lowest.min(vtt_min),
+                        vtt_highest.max(vtt_min),
+                    )
+                },
+            );
+
+        // The size of each bucket in nWitWu (nano wits per weight unit)
+        let drt_buckets_size = (drt_highest_absolute - drt_lowest_absolute) / buckets_count;
+        let vtt_buckets_size = (vtt_highest_absolute - vtt_lowest_absolute) / buckets_count;
+
+        // Now we are ready to map priorities to buckets and insert the bucket numbers into the
+        // lossy counter.
+        for Priorities {
+            drt_highest,
+            drt_lowest,
+            vtt_highest,
+            vtt_lowest,
+        } in priorities
+        {
+            // This calculates the buckets in which the lowest values should be inserted.
+            let drt_bucket = ((drt_lowest.unwrap_or(*drt_highest).as_f64() - drt_lowest_absolute)
+                / drt_buckets_size)
+                .round() as u64;
+            let vtt_bucket = ((vtt_lowest.unwrap_or(*vtt_highest).as_f64() - vtt_lowest_absolute)
+                / vtt_buckets_size)
+                .round() as u64;
```

Comment:
There is an edge case here when `vtt_buckets_size` is zero, which can happen when `vtt_highest_absolute == vtt_lowest_absolute`. That results in `0.0/0.0` which is `NaN`, however after casting to `u64` it turns into `0`, and that's the correct behavior. In that case we don't even need to iterate over the priorities, because the estimated fee will always be `vtt_lowest_absolute + 0 * 0`, but I guess handling that edge case explicitly would result in more complex code, so it's fine this way.